quickwit/quickwit-janitor/src/garbage_collection.rs

```diff
@@ -243,100 +235,104 @@ async fn delete_splits_marked_for_deletion(
 /// * `metastore` - The metastore managing the target index.
 /// * `splits`  - The list of splits to delete.
 /// * `ctx_opt` - A context for reporting progress (only useful within quickwit actor).
-pub async fn delete_splits_with_files(
+pub async fn delete_splits_from_storage_and_metastore(
     index_uid: IndexUid,
     storage: Arc<dyn Storage>,
     metastore: Arc<dyn Metastore>,
     splits: Vec<SplitMetadata>,
     ctx_opt: Option<&ActorContext<GarbageCollector>>,
-) -> anyhow::Result<Vec<FileEntry>, SplitDeletionError> {
-    let mut paths_to_splits = HashMap::with_capacity(splits.len());
+) -> anyhow::Result<Vec<SplitInfo>, DeleteSplitsError> {
+    let mut split_infos: HashMap<PathBuf, SplitInfo> = HashMap::with_capacity(splits.len());
 
     for split in splits {
-        let file_entry = FileEntry::from(&split);
-        let split_filename = quickwit_common::split_file(split.split_id());
-        let split_path = Path::new(&split_filename);
-
-        paths_to_splits.insert(
-            split_path.to_path_buf(),
-            (split.split_id().to_string(), file_entry),
-        );
+        let split_info = split.as_split_info();
+        split_infos.insert(split_info.file_name.clone(), split_info);
     }
-
-    let paths = paths_to_splits
+    let split_paths = split_infos
         .keys()
-        .map(|key| key.as_path())
+        .map(|split_path_buf| split_path_buf.as_path())
         .collect::<Vec<&Path>>();
-    let delete_result = storage.bulk_delete(&paths).await;
+    let delete_result = protect_future(ctx_opt, storage.bulk_delete(&split_paths)).await;
 
     if let Some(ctx) = ctx_opt {
         ctx.record_progress();
     }
+    let mut storage_error: Option<BulkDeleteError> = None;
+    let mut storage_failures = Vec::new();
+
+    if let Err(bulk_delete_error) = delete_result {
+        storage_failures = bulk_delete_error
+            .failures
+            .keys()
+            .chain(&bulk_delete_error.unattempted)
+            .flat_map(|split_path| split_infos.remove(split_path))
+            .collect::<Vec<_>>();
+        error!(
+            error=?bulk_delete_error.error,
+            index_id=index_uid.index_id(),
+            "Failed to delete split files {:?} from storage.",
+            PrettySample::new(&storage_failures, 5),
+        );
+        storage_error = Some(bulk_delete_error);
+    }
+    if !split_infos.is_empty() {
+        let split_ids: Vec<&str> = split_infos
+            .values()
+            .map(|split_info| split_info.split_id.as_str())
+            .collect();
```

Comment:
I'm a bit worried a bug in `bulk_delete` could not add a split in any list (`failures`, `unattempted`, `success`), and we would consider it a success for that split. I think it would be more defensive to use `bulk_delete_error.success` if `bulk_delete` failed, and only consider everything in `split_infos` is okay when `delete_result` is `Ok(_)`

Reply:
I checked the storage code before going with this strategy and we seemed to be doing the right, but I'm ok the more belt and suspenders approach that you describe. The ` .flat_map(|split_path| split_infos.remove(split_path))` was bad anyway. That is what you get for coding on Sunday mornings at 6 am.